cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -242,90 +296,74 @@ static __device__ int gpuUpdateValidityAndRowIndicesNested(
+        int const vindex     = value_count + thread_value_count;  // absolute input value index
+        int const bit_offset = (valid_map_offset + vindex + write_start) -
+                               first_row;  // absolute bit offset into the output validity map
+        int const write_end =
+          cudf::detail::warp_size - __clz(in_write_row_bounds);  // last bit in the warp to store
+        int const bit_count = write_end - write_start;
+        warp_null_count     = bit_count - __popc(warp_validity_mask >> write_start);
+
+        store_validity(bit_offset, ni.valid_map, warp_validity_mask >> write_start, bit_count);
       }
 
+      // sum null counts. we have to do it this way instead of just incrementing by (value_count -
+      // valid_count) because valid_count also includes rows that potentially start before our row
+      // bounds. if we could come up with a way to clean that up, we could remove this and just
+      // compute it directly at the end of the kernel.
+      size_type const block_null_count =
+        cudf::detail::single_lane_block_sum_reduce<decode_block_size, 0>(warp_null_count);
+      if (t == 0) { ni.null_count += block_null_count; }
+
       // if this is valid and we're at the leaf, output dst_pos
-      __syncthreads();  // handle modification of ni.value_count from below
-      if (is_valid && d_idx == max_depth) {
-        // for non-list types, the value count is always the same across
-        int const dst_pos = (value_count + thread_value_count) - 1;
-        int const src_pos = (ni.valid_count + thread_valid_count) - 1;
-        sb->nz_idx[rolling_index<state_buf::nz_buf_size>(src_pos)] = dst_pos;
+      if (d_idx == max_depth) {
+        if (is_valid) {
+          // for non-list types, the value count is always the same across
+          int const dst_pos = value_count + thread_value_count;
+          int const src_pos = max_depth_valid_count + thread_valid_count;
+          sb->nz_idx[rolling_index<state_buf::nz_buf_size>(src_pos)] = dst_pos;
+        }
+        // update stuff
+        max_depth_valid_count += block_valid_count;
       }
-      __syncthreads();  // handle modification of ni.value_count from below
 
-      // update stuff
-      if (t == 0) { ni.valid_count += block_valid_count; }
-    }
+    }  // end depth loop
 
     value_count += block_value_count;
-  }
+  }  // end loop
 
   if (t == 0) {
     // update valid value count for decoding and total # of values we've processed
-    s->nz_count          = s->nesting_info[max_depth].valid_count;
-    s->input_value_count = value_count;
-    s->input_row_count   = value_count;
+    max_depth_ni.valid_count = max_depth_valid_count;
+    s->nz_count              = max_depth_valid_count;
+    s->input_value_count     = value_count;
+    s->input_row_count       = value_count;
   }
 
-  __syncthreads();
-  return s->nesting_info[max_depth].valid_count;
+  return max_depth_valid_count;
```

Comment:
Is there any way for this to be wrong for some threads?  It seems like every thread will have participated in everything they need so that this is correct, but I wanted to verify.  I guess they must have since previously we were doing a full `__syncthreads()`.

Reply:
We call __syncthreads() immediately after calling this function in the main while loop in gpuDecodePageDataGeneric().  So we can put them one place or the other (your preference is fine), but having them in both didn't add anything.